Update and fix div/shift checks

MartinNowack opened the pull request:
This PR provides a couple of goodies:
* Use `llvm::Builder` instead of `*Inst::create()`:
  - manages metadata automatically (e.g. debug, TBAA, ...)
  - does optimisation like code folding with newly added instructions (no `bitcasts` if none needed)
* Fixes an issue introduced by the #868: checks for single instructions can be inserted multiple times: avoid this by flagging an instruction as instrumented
 - Add basic support to annotate code with KLEE
* Optimisation: Do not instrument instructions that do not need instrumentation (e.g. div by non-zero constant)
* Add tests to check the optimised and non-optimised case

Update with C++11 and clang-format.

MartinNowack commented:
@jbuening Thanks a lot for the great review and suggestions! I updated the code appropriately.
This should be ready now as well.

MartinNowack commented:
@jbuening Thanks for spotting those mistakes. Fixed! Any other suggestion/updates are welcome.
Otherwise we are ready to go.

jbuening commented:
@MartinNowack Thanks again! At least from my side, there are no further suggestions on this PR.

ccadar commented:
Hi @MartinNowack , thanks for the patch, these are nice changes and the code looks much cleaner.
I have three comments:
1) `coOp->getZExtValue() <= typeWidth` should be `coOp->getZExtValue() < typeWidth`
2) Could you add some comments to the tests as to what they are testing?  I'm still a bit confused as to what assumptions you make about `--optimize` for instance
3) I didn't like that https://github.com/klee/klee/pull/945/commits/4a8a60990f0ba5a290c48042dc4d04b9a31a02d6 is not self-contained, because you never update `divInstruction` 

Thanks again!

MartinNowack commented:
@ccadar Thanks for the review, I addressed all your issues except number `3.` The referenced commit is about the div checks not directly about `divInstructions` can you elaborate that point more? Thanks!

ccadar commented:
Thanks, @MartinNowack .  Let's not worry about 3, it's minor.

MartinNowack commented:
@ccadar In case you have some time, please merge.

ccadar commented:
Thanks again, @MartinNowack , merging!
